File: Practica1/Ejercicio4/src/practica1ejercicio4/CadenaFiltros.java
package practica1ejercicio4;

import java.util.ArrayList;
import java.util.List;

public class CadenaFiltros implements Filtro {
    private List<Filtro> filtros = new ArrayList<>();
    private Objetivo correoFiltro;

    public void addFiltro(Filtro filtro) {
        filtros.add(filtro);
    }

    public void setObjetivo(Objetivo correoFiltro) {
        this.correoFiltro = correoFiltro;
    }

    @Override
    public boolean ejecutar(Correo correo) {
        for (Filtro filtro : filtros) {
            if (!filtro.ejecutar(correo)) {
                System.out.println("Correo rechazado: " + correo.getEmail());
                return false;
            }
        }

        System.out.println("Correo aceptado: " + correo.getEmail());
        correoFiltro.string(correo);
        return true;
    }
}

File: Practica2/ejercicio_grupal1/lib/CadenaFiltros.dart
import 'Filtro.dart';
import 'Objetivo.dart';
import 'Correo.dart';

class CadenaFiltros implements Filtro {
  final List<Filtro> filtros = [];
  late Objetivo correoFiltro;

  void addFiltro(Filtro filtro) {
    filtros.add(filtro);
  }

  void setObjetivo(Objetivo correoFiltro) {
    this.correoFiltro = correoFiltro;
  }

  @override
  MapEntry<bool,String> ejecutar(Correo correo) {
    MapEntry<bool,String> output;

    for (var filtro in filtros) {
      output = filtro.ejecutar(correo);
      if (!output.key) { // En caso de que haya algún correo rechazado
        print("Correo rechazado: ${correo.email}");
        return output;
      }
    }

    print("Correo aceptado: ${correo.email}");
    correoFiltro.string(correo);
    output = MapEntry(true, "Correo y contraseña aceptados");
    return output;
  }
}